### ISSUE TITLE
Nside=4096 is prohibitively expensive

### ISSUE THREAD

**Issue by keskitalo** [healpy/pysm#30]
Trying to run LAT LF simulation at `Nside=4096`. There are 222 detectors. I can only allocate two MPI processes on each KNL node not to exhaust the memory during PySM call. I split the communicator into 256 groups, each with one node and two processes. There are two rank communicators with 256 MPI processes, each trying to simulate 111 detectors. The job did not complete in two hours. The medium and high frequency tubes have 10X as many detectors.

**Comment by zonca** [comment]
I can run some tests to check if this is expected,
do you have a feeling on how this compares to my benchmark at https://github.com/healpy/pysm/issues/25#issuecomment-500657096?

**Comment by zonca** [comment]
how many threads are you using per MPI process?

**Comment by keskitalo** [comment]
That comment indicates 40 MPI tasks processing a channel every 60 seconds. Naively assuming perfect scaling, 256 MPI tasks would do the same in 10 seconds or 111 detectors in less than 20 minutes. Of course the KNL cores are slower but we have 32 OpenMP threads for each MPI task.

**Comment by zonca** [comment]
I started with a scaled down test on Popeye (48 core Intel skylake), basically 1/10 of the job:

13 nodes, 2 MPI processes per node, 24 OpenMP/Numba
* Initialize sky: 143s 
* Bandpass integration per channel (10 points like TOAST now): 80s the first channel (loading CIB maps), 6s the others
* Smoothing: 7s
* Memory (with memreport) 9.5GB after initialization, 10 GB at the end

So just spreading maps over 26 MPI processes keeps memory consumption at ~10GB, is it possible that there are memory issues in other parts of the PySM operator? i.e. the map redistribution?

**Comment by keskitalo** [comment]
Thanks Andrea. This is really helpful. I cannot say if the memory consumption is driven by PySM or other parts of `OpSimPysm`. What happens if you try to run more than 2 processes per node?

`memreport` gives you the instantaneous memory usage but will not track the high water mark. Unless you are also calling it during PySM it won't be very helpful.

**Comment by zonca** [comment]
had the same question:

If I run 4 processes per node, same total number of nodes,
Initialize sky: 148s
Bandpass integration per channel (10 points like TOAST now): 80s the first channel (loading and caching CIB maps), 4s the others
Smoothing: 7s
Memory (with memreport) 9.35GB after initialization, 10.5 GB at the end
Total time for 10 channels: 6 minutes

`memreport` is per node, right?

**Comment by zonca** [comment]
If I run 6 processes per node, 8 threads, same total number of nodes,
Initialize sky: 137s
Bandpass integration per channel (10 points like TOAST now): 75s the first channel (loading and caching CIB maps), 2s the others
Smoothing: 6s
Memory (with memreport) 7.7GB after initialization, 11.6 GB at the end
Total time for 10 channels: 5 minutes

**Comment by keskitalo** [comment]
Correct: `memreport` is by node.  It is very impressive that you get the same walltime with 4 and 6 processes per node! Can you fit even more processes per node? Are those numbers the available memory on the node or the total allocated?

**Comment by keskitalo** [comment]
Wow. That is a lot of memory. Without calling `memreport` inside PySM those numbers do not mean much.

**Comment by keskitalo** [comment]
Comparing the 4x12 to 2x24 the memory consumption is 70% higher. 6x8 needs 130% more memory. Those are notable increases. If the maps were distributed evenly across the communicator, we would not expect any increase, would we?

Perhaps the map distribution in PySM is set up to broadcast the high resolution maps across the communicator before each process extracts its part of the map? If that were the case, it would drive up the memory high water mark and cause an unnecessary out-of-memory. Buffering the broadcast or using `scatterv` would avoid this.

This all seems compatible with my observation that at `nside=4096` we cannot fit more than 2 MPI tasks on each KNL node. What happens when we go to `nside=8192`? TOAST includes a significant amount of Python glue that does not thread at all so limiting the number of MPI tasks like this is problematic.

The actual execution times you get are much better than my out-of-time test would have suggested. I wonder if `OpSimPysm` includes costly operations not included in your test. We probably need a test that runs the entire operator, not just PySM.

**Comment by zonca** [comment]
yes, correct, I broadcast the full map. I looked into using `scatterv`, but each process has rings which are not consecutive, and maps are often 2D, so it is going to be very complicated.

Also, in the worst case of IQU map at 4096, we are paying a temporary increase of 2.3 GB of memory per MPI process, so let's say you want to run 4 MPI processes per node, that would be ~10 GB. Or if we compare running 2 MPI tasks or 4 MPI tasks, in the second case you pay an extra 4.6 GB per node temporarily. It doesn't seem too bad. I am more worried about other parts of the PySM Operator.

I'll try to install TOAST on Popeye and see if I can run a test of the full `OpSimPysm`

**Comment by keskitalo** [comment]
So that is 40GB extra memory overhead for `Nside=8192` and just 4 MPI processes per node? I don't think that is the best approach. If `scatterv` is not appropriate, then you can broadcast the map in 100MB pieces.

**Comment by zonca** [comment]
consider that this price is paid only:

* at initialization for most components
* during 1st channel bandpass integration for CIB (before smoothing)

so there are no timelines yet, so do you plan to have less than 40GB of timelines per node?

**Comment by keskitalo** [comment]
At this point the pointing has to be expanded already, so even if the signal doesn't yet exist, pixel numbers and pointing weights are in memory. `OpSimPySM` may also be adding to an existing signal so you cannot assume that memory to be available as workspace.

Furthermore, we need to have 8 or 16 MPI tasks per node, not 2 or 4.

**Comment by zonca** [comment]
could we use MPI shared memory?

**Comment by keskitalo** [comment]
That would be a step in the right direction. It still means that `nside=8192` would require 10GB of workspace but at least it would not scale with the number of MPI tasks per node.

**Comment by zonca** [comment]
@keskitalo is the PySM call going out of memory in the initialization phase or during integration of the first channel or later on?

**Comment by keskitalo** [comment]
I don't think the log gave that information. The program termination under OOM is immediate. I'll check once NERSC is up again.

**Comment by zonca** [comment]
See mpi shared memory example in TOAST: https://github.com/hpc4cmb/toast/blob/master/src/toast/mpi.py#L532

**Comment by zonca** [comment]
I think the ideal solution would be to have input files in HDF5 and read them directly with a MPI-IO call into local buffers in each MPI process. @keskitalo would this be an option? Would this solution scale to the jobs you need for the current SO simulations?

**Comment by keskitalo** [comment]
So every process would read the same file? I suppose it works until it doesn't. You should make sure to open the files in read-only mode or otherwise the access is not parallel. We usually prefer read and broadcast because it will scale to very large concurrencies.

Are you planning on adding HDF5 support to Healpy? It is a lot of boilerplate to handle otherwise.

**Comment by zonca** [comment]
Debugging the memory issue with this configuration (after having implemented the mpi_shared feature in #32):

The low frequency LAT simulation only has 222 detectors. The nominal case runs with 32 nodes. I allocated 128 nodes for the realistic case, organized into groups of 16 nodes (8 MPI procsses per node). This job fails with out-of-memory during OpSimPySM.

Submitted equivalent job on Popeye to understand memory usage.

**Comment by zonca** [comment]
ran on Cori the equivalent of 1 group 16 nodes 8 MPI processes per node, see `memreport` below:

```
Memory usage                                                                                                                                      total :   94.250 GB  <   94.250 +-    0.000 GB  <   94.250 GB
   available :   71.759 GB  <   72.441 +-    0.188 GB  <   72.663 GB
   available :   72.415 GB  <   72.529 +-    0.058 GB  <   72.663 GB (after GC)                                                            
     percent :   22.900 %   <   23.100 +-    0.198 %   <   23.900 %
     percent :   22.900 %   <   23.000 +-    0.079 %   <   23.200 %  (after GC)
        used :    6.147 GB  <    6.356 +-    0.188 GB  <    7.027 GB                                                                       
        used :    6.146 GB  <    6.270 +-    0.058 GB  <    6.378 GB (after GC)
        free :   72.016 GB  <   72.817 +-    0.198 GB  <   73.052 GB
        free :   72.672 GB  <   72.912 +-    0.079 GB  <   73.053 GB (after GC)                                                            
      active :    4.141 GB  <    4.233 +-    0.181 GB  <    4.889 GB                                                                             active :    4.141 GB  <    4.155 +-    0.023 GB  <    4.242 GB (after GC)
    inactive :   13.991 GB  <   14.005 +-    0.046 GB  <   14.154 GB                                                                       
     buffers :    9.590 MB  <   10.043 +-    7.348 MB  <   40.336 MB                                                                             cached :   15.042 GB  <   15.051 +-    0.046 GB  <   15.200 GB
      shared :   14.621 GB  <   14.629 +-    0.006 GB  <   14.638 GB                                                                       
        slab :    1.172 GB  <    1.181 +-    0.007 GB  <    1.198 GB 
```

**Comment by keskitalo** [comment]
Are you calling `memreport` at the time of the highest memory requirement or after the operator finishes? Recall that `memreport` does not track the high watermark but only the current memory.

**Comment by zonca** [comment]
in my tests on Popeye `memreport` (inside the Operator) was pretty close with what I got with `seff`. @keskitalo what do you use to get high watermark on Cori?

**Comment by keskitalo** [comment]
I just call `memreport` where the memory consumption is expected to be the highest.

**Comment by zonca** [comment]
Asked NERSC, they don't have any suggestion on how to get the high watermark properly.
So I created 4GB timelines for each process,
so now even before PySM we are at:

```
***** Created fake timelines of 4.00 GB per process
Walltime elapsed since the beginning 6.9 s
Memory usage 
       total :   94.250 GB  <   94.250 +-    0.000 GB  <   94.250 GB
   available :   57.072 GB  <   57.503 +-    0.166 GB  <   57.995 GB
   available :   57.072 GB  <   57.428 +-    0.119 GB  <   57.584 GB (after GC)
     percent :   38.500 %   <   39.000 +-    0.176 %   <   39.400 % 
     percent :   38.900 %   <   39.100 +-    0.117 %   <   39.400 %  (after GC)
        used :   35.075 GB  <   35.567 +-    0.168 GB  <   36.011 GB
        used :   35.486 GB  <   35.643 +-    0.120 GB  <   36.011 GB (after GC)
        free :   57.483 GB  <   57.911 +-    0.166 GB  <   58.402 GB
        free :   57.483 GB  <   57.839 +-    0.119 GB  <   57.991 GB (after GC)
      active :   32.367 GB  <   32.796 +-    0.113 GB  <   32.941 GB
      active :   32.777 GB  <   32.867 +-    0.028 GB  <   32.941 GB (after GC)
    inactive :  530.578 MB  <  540.186 +-   14.302 MB  <  589.172 MB
     buffers :    8.234 MB  <    8.707 +-    0.320 MB  <    9.312 MB
      cached :  766.215 MB  <  774.553 +-   14.428 MB  <  825.289 MB
      shared :  379.680 MB  <  388.158 +-   14.145 MB  <  436.152 MB
        slab :    1.137 GB  <    1.156 +-    0.012 GB  <    1.187 GB
```

at the end:

```
***** Ran smoothing channel 9
Walltime elapsed since the beginning 1428.4 s
Memory usage 
       total :   94.250 GB  <   94.250 +-    0.000 GB  <   94.250 GB
   available :   39.601 GB  <   40.327 +-    0.205 GB  <   40.539 GB
   available :   40.177 GB  <   40.445 +-    0.101 GB  <   40.544 GB (after GC)
     percent :   57.000 %   <   57.200 +-    0.219 %   <   58.000 % 
     percent :   57.000 %   <   57.100 +-    0.116 %   <   57.400 %  (after GC)
        used :   38.268 GB  <   38.470 +-    0.206 GB  <   39.200 GB
        used :   38.263 GB  <   38.353 +-    0.103 GB  <   38.634 GB (after GC)
        free :   39.989 GB  <   40.716 +-    0.205 GB  <   40.927 GB
        free :   40.566 GB  <   40.832 +-    0.100 GB  <   40.931 GB (after GC)
      active :   36.141 GB  <   36.246 +-    0.180 GB  <   36.906 GB
      active :   36.141 GB  <   36.174 +-    0.025 GB  <   36.246 GB (after GC)
    inactive :   13.996 GB  <   14.005 +-    0.014 GB  <   14.054 GB
     buffers :    9.504 MB  <    9.965 +-    0.309 MB  <   10.516 MB
      cached :   15.040 GB  <   15.049 +-    0.014 GB  <   15.097 GB
      shared :   14.621 GB  <   14.629 +-    0.014 GB  <   14.676 GB
        slab :    1.172 GB  <    1.191 +-    0.012 GB  <    1.222 GB
```

Now trying with 8GB per MPI process.

**Comment by zonca** [comment]
works fine also with 8GB timelines per MPI process:

```
***** Created fake timelines of 8.00 GB per process
Walltime elapsed since the beginning 11.5 s
Memory usage
       total :   94.250 GB  <   94.250 +-    0.000 GB  <   94.250 GB
   available :   25.311 GB  <   25.609 +-    0.225 GB  <   26.557 GB
   available :   25.311 GB  <   25.472 +-    0.108 GB  <   26.098 GB (after GC)
     percent :   71.800 %   <   72.800 +-    0.244 %   <   73.100 %
     percent :   72.300 %   <   73.000 +-    0.115 %   <   73.100 %  (after GC)
        used :   66.527 GB  <   67.475 +-    0.225 GB  <   67.763 GB
        used :   66.985 GB  <   67.611 +-    0.107 GB  <   67.763 GB (after GC)
        free :   25.717 GB  <   26.020 +-    0.226 GB  <   26.970 GB
        free :   25.717 GB  <   25.883 +-    0.109 GB  <   26.512 GB (after GC)
      active :   63.788 GB  <   64.733 +-    0.224 GB  <   64.968 GB
      active :   64.243 GB  <   64.884 +-    0.106 GB  <   64.968 GB (after GC)
    inactive :  531.605 MB  <  533.812 +-    1.932 MB  <  539.012 MB
     buffers :    8.695 MB  <    8.869 +-    0.197 MB  <    9.523 MB
      cached :  762.945 MB  <  765.391 +-    3.761 MB  <  780.059 MB
      shared :  379.426 MB  <  379.451 +-    1.207 MB  <  384.445 MB
        slab :    1.175 GB  <    1.179 +-    0.002 GB  <    1.183 GB
```

at the end:

```
***** Ran smoothing channel 9
Walltime elapsed since the beginning 1419.1 s
Memory usage
       total :   94.250 GB  <   94.250 +-    0.000 GB  <   94.250 GB
   available :    7.697 GB  <    8.354 +-    0.186 GB  <    8.496 GB
   available :    8.340 GB  <    8.446 +-    0.038 GB  <    8.496 GB (after GC)
     percent :   91.000 %   <   91.100 +-    0.200 %   <   91.800 %
     percent :   91.000 %   <   91.000 +-    0.061 %   <   91.200 %  (after GC)
        used :   70.315 GB  <   70.456 +-    0.186 GB  <   71.112 GB
        used :   70.315 GB  <   70.366 +-    0.036 GB  <   70.459 GB (after GC)
        free :    8.086 GB  <    8.744 +-    0.186 GB  <    8.884 GB
        free :    8.723 GB  <    8.836 +-    0.039 GB  <    8.884 GB (after GC)
      active :   68.167 GB  <   68.290 +-    0.185 GB  <   68.952 GB
      active :   68.167 GB  <   68.210 +-    0.030 GB  <   68.271 GB (after GC)
    inactive :   13.991 GB  <   14.000 +-    0.006 GB  <   14.017 GB
     buffers :    9.957 MB  <   10.131 +-    0.197 MB  <   10.785 MB
      cached :   15.038 GB  <   15.041 +-    0.004 GB  <   15.057 GB
      shared :   14.621 GB  <   14.621 +-    0.001 GB  <   14.626 GB
        slab :    1.210 GB  <    1.214 +-    0.002 GB  <    1.218 GB
```

**Comment by zonca** [comment]
10GB failed with OOM

**Comment by zonca** [comment]
Looking again to this I realized that the MPI shared memory goes into `shared` and that is not properly cleared by Python. So I am now calling `win.Free()` explicitly and memory usage went down.
Now also 10 GB fake timelines per node work fine:

```
***** Ran smoothing channel 9
Walltime elapsed since the beginning 1449.1 s
Memory usage
       total :   94.250 GB  <   94.250 +-    0.000 GB  <   94.250 GB
   available :    5.609 GB  <    6.718 +-    0.273 GB  <    6.950 GB
   available :    6.005 GB  <    6.819 +-    0.214 GB  <    6.957 GB (after GC)
     percent :   92.600 %   <   92.900 +-    0.285 %   <   94.000 %
     percent :   92.600 %   <   92.800 +-    0.228 %   <   93.600 %  (after GC)
        used :   86.117 GB  <   86.348 +-    0.269 GB  <   87.435 GB
        used :   86.110 GB  <   86.245 +-    0.208 GB  <   87.039 GB (after GC)
        free :    6.010 GB  <    7.123 +-    0.274 GB  <    7.356 GB
        free :    6.406 GB  <    7.224 +-    0.215 GB  <    7.362 GB (after GC)
      active :   83.297 GB  <   83.411 +-    0.190 GB  <   84.059 GB
    inactive :  544.371 MB  <  555.404 +-    5.817 MB  <  571.734 MB
     buffers :    9.344 MB  <   10.277 +-    0.418 MB  <   10.945 MB
      cached :  781.066 MB  <  787.771 +-    7.591 MB  <  814.453 MB
      shared :  382.914 MB  <  388.900 +-    5.399 MB  <  406.613 MB
        slab :    1.123 GB  <    1.134 +-    0.012 GB  <    1.164 GB
```

See how shared memory is reduced now.
I'm closing this issue, reopen if needed.
